docs: add comprehensive firewall rules documentation (#1660)
kubernetes-sigs/cluster-api-provider-gcp: docs/book/src/topics/firewall-rules.md

# Firewall Rules

> Cluster API Provider GCP allows you to configure GCP VPC firewall rules for your clusters through the `GCPCluster` resource. This feature provides fine-grained control over network access to your cluster infrastructure.

Reviewer: What about `GCPManagedCluster` (GKE)? I think firewall spec is also supported there, right?

> Cluster API Provider GCP allows you to configure GCP VPC firewall rules for your clusters through the `GCPCluster` resource. This feature provides fine-grained control over network access to your cluster infrastructure.

Reviewer: ```suggestion
Cluster API Provider GCP (CAPG) allows you to configure GCP VPC firewall rules for your clusters through the `GCPCluster` resource. This feature provides fine-grained control over network access to your cluster infrastructure.
```

nit: Let's add this so we can use the word `CAPG` later on :D

## Overview

Firewall rules are configured through the `network.firewall` field in the `GCPCluster` spec. The firewall configuration supports:

- **Default rule management**: Control whether the provider creates default firewall rules
- **Custom firewall rules**: Define additional firewall rules to meet your specific security requirements

## Default Firewall Rules

By default, the provider creates two firewall rules to enable cluster functionality:

1. **Health Check Rule** (`allow-<cluster-name>-healthchecks`):

>    - Allows TCP traffic on port 6443 from GCP health check IP ranges

Reviewer: ```suggestion
   - Allows TCP ingress traffic on port 6443 from GCP health check IP ranges
```

nit: let's explicitly state that the rule is `INGRESS`.

   - Source ranges: `35.191.0.0/16`, `130.211.0.0/22`
   - Target: control-plane nodes

2. **Cluster Internal Rule** (`allow-<cluster-name>-cluster`):

>    - Allows all traffic between cluster nodes

Reviewer: ```suggestion
   - Allows all ingress traffic between cluster nodes
```

nit: let's explicitly state that the rule is `INGRESS`.

   - Source/Target: control-plane and worker nodes

### Managing Default Rules

You can control the creation of default firewall rules using the `defaultRulesManagement` field:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: GCPCluster
metadata:
  name: my-cluster
spec:
  network:
    firewall:
      defaultRulesManagement: "Managed"  # or "Unmanaged"
```

**Values:**
- `Managed` (default): The controller creates and manages default firewall rules

> - `Unmanaged`: The controller does not create or modify default firewall rules

Reviewer: ```suggestion
- `Managed` (default): CAPG creates and manages default firewall rules
- `Unmanaged`: CAPG does not create or manage default firewall rules
```

nit: let's use `CAPG` and similar verb between the 2 modes for consistency?

**Important Notes:**

> - Changing from `Managed` to `Unmanaged` after rules are created will not delete existing rules

Reviewer: ```suggestion
- CAPG allows switching from `Managed` to `Unmanaged` and vice versa. However, changing from `Managed` to `Unmanaged` after rules are created will not delete existing rules. Switching from `Unmanaged` to `Managed` will cause CAPG to create the default rules if they don’t already exist
```

Let's state that "switching" is allowed before mentioning the caveats?

Additionally, what happens if changing from `Unmanaged` to `Managed`? CAPG will create the default rules if not exist, right?

> - `defaultRulesManagement` has no effect when using a shared VPC (HostProject)

Reviewer: 💡 The user may wonder what CAPG behaviours will be when using a shared VPC. How about calling out that:

>  When using a shared VPC (`HostProject`), the controller will not create, modify, or delete any firewall rules (default or custom).

      defaultRulesManagement: "Managed"
      firewallRules:
        - name: "custom-ingress-rule"
          description: "Allow SSH and custom application traffic"
          direction: "Ingress"
          priority: 1000
          allowed:
            - IPProtocol: "TCP"
              ports:
                - "22"
                - "8080"
                - "8443"
          sourceRanges:
            - "10.0.0.0/8"
            - "172.16.0.0/12"
          targetTags:
            - "web-servers"
```

### FirewallRule Fields

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `name` | string | Optional | Rule name (1-63 chars, must match `[a-z]([-a-z0-9]*[a-z0-9])?`). If not prefixed with cluster name, it will be prepended automatically. |
| `description` | string | Optional | Description of the rule (max 2000 chars). Defaults to "Created by Cluster API GCP Provider". |
| `direction` | string | Optional | Traffic direction: `Ingress` (default) or `Egress`. |
| `priority` | integer | Optional | Rule priority (1-65535). Lower values = higher priority. Default: 1000. |
| `allowed` | []FirewallDescriptor | Optional | List of ALLOW rules (max 1024). |

> | `denied` | []FirewallDescriptor | Optional | List of DENY rules (max 1024). DENY rules take precedence over ALLOW rules with equal priority. |

Reviewer: ```suggestion
| `denied` | []FirewallDescriptor | Optional | List of DENY rules (max 1024). DENY rules take precedence over ALLOW rules if they have equal priority. |
```

nit: just a little clearer :D

| `sourceRanges` | []string | Optional | Source IP ranges in CIDR format (max 1024). Supports IPv4 and IPv6. |

> | `sourceTags` | []string | Optional | Source instance tags (max 30, 1-63 chars each). Only applies to traffic between instances in the same VPC. |

Reviewer: ```suggestion
| `sourceTags` | []string | Optional | Source instance tags (max 30, 1-63 chars each). Only applies to traffic between instances in the same VPC. Not valid for Egress rules. |
```

This field is not valid for Egress, right?

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/23049b61863d1608729391f70f5a2dfb4257472b/api/v1beta1/types.go#L188-L189

> | `destinationRanges` | []string | Optional | Destination IP ranges in CIDR format (max 1024). Only valid for Egress rules. |

Reviewer: I notice that the `destinationRanges` is defined, but unused by CAPG.  Is this a bug (perhaps accidentially missed)?

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/23049b61863d1608729391f70f5a2dfb4257472b/cloud/scope/firewall.go#L96-L108

              ports:
                - "80"
                - "443"
          sourceRanges:
            - "0.0.0.0/0"
          targetTags:
            - "web-servers"
```

### Example 3: Deny Rule with Higher Priority

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: GCPCluster
metadata:
  name: my-cluster
spec:
  network:
    firewall:
      firewallRules:
        - name: "deny-telnet"
          description: "Block telnet traffic"
          direction: "Ingress"
          priority: 500
          denied:
            - IPProtocol: "TCP"
              ports:
                - "23"
          sourceRanges:
            - "0.0.0.0/0"
```

### Example 4: Egress Rule for External Services

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: GCPCluster
metadata:
  name: my-cluster
spec:
  network:
    firewall:
      firewallRules:
        - name: "allow-external-api"
          description: "Allow egress to external API"
          direction: "Egress"
          priority: 1000
          allowed:
            - IPProtocol: "TCP"
              ports:
                - "443"
          destinationRanges:
            - "198.51.100.0/24"
          sourceTags:

>             - "api-client"

Reviewer: Egress rule cannot use `sourceTags` right?

  project: my-gcp-project
  region: us-central1
  network:
    name: production-network
    firewall:
      # Manage default health check and cluster internal rules
      defaultRulesManagement: "Managed"
      # Add custom rules
      firewallRules:
        # Allow monitoring from Prometheus
        - name: "monitoring"
          description: "Allow Prometheus scraping"
          direction: "Ingress"
          priority: 900
          allowed:
            - IPProtocol: "TCP"
              ports:
                - "9090"
                - "9100"
          sourceRanges:
            - "10.128.0.0/16"
          targetTags:
            - "prometheus-target"
        # Allow database access from application tier
        - name: "database-access"
          description: "Allow app to database communication"
          direction: "Ingress"
          priority: 1000
          allowed:
            - IPProtocol: "TCP"
              ports:
                - "5432"
                - "3306"
          sourceTags:
            - "app-tier"
          targetTags:
            - "database-tier"
```

## Best Practices

1. **Use Descriptive Names**: Choose meaningful names that describe the rule's purpose

> 2. **Set Appropriate Priorities**: Use lower values for more critical security rules

Reviewer: ```suggestion
2. **Set Appropriate Priorities**: Use lower values (higher priority) for more critical security rules
```

3. **Minimize Source Ranges**: Restrict access to only necessary IP ranges
4. **Use Tags Strategically**: Leverage instance tags for flexible rule targeting
5. **Document Rules**: Always include a description explaining the rule's purpose
6. **Test Before Production**: Verify firewall rules in a development environment first
7. **Avoid Priority 65535**: GCP reserves this priority for implied rules
8. **Consider DENY Rules**: Use DENY rules for explicit blocking with higher priority

## Limitations

- Maximum 50 custom firewall rules per cluster

> - Maximum 1024 allowed/denied descriptors per rule

Reviewer: nit: we probably can remove this section since they are mostly mention in the table above :D

# Firewall Rules

Cluster API Provider GCP (CAPG) allows you to configure GCP VPC firewall rules for your clusters through the `GCPCluster` and `GCPManagedCluster` (GKE) resources. This feature provides fine-grained control over network access to your cluster infrastructure.

## Overview

> Firewall rules are configured through the `network.firewall` field in the `GCPCluster` or `GCPManagedCluster` spec. Both resource types share the same network specification, so firewall configuration works identically for self-managed and GKE clusters. The firewall configuration supports:

Reviewer: 🤦 oh whoops, the firewall field is defined on `GCPManagedCluster` (GKE), but never really got reconciled/created?

I'm looking at these registered reconciler. IIUC, firewall specs have no effect.

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/95b916103f0b609a8556b2f3345325fbaa04771c/exp/controllers/gcpmanagedcluster_controller.go#L198-L201

- **Default rule management**: Control whether the provider creates default firewall rules

> - **Custom firewall rules**: Define additional firewall rules to meet your specific security requirements

Reviewer: We should mention that:
- CAPG doesn't actively update the firewall rules to match their current (if updated) specs after creation
- CAPG doesn't remove firewall rules if they are not in the list anymore.

AFAIK, CAPG only creates rules and deletes the rules only when cluster is deleted, right?

- When using a shared VPC (`HostProject`), CAPG will not create, modify, or delete any firewall rules (default or custom)

## Custom Firewall Rules

You can define up to 50 additional firewall rules using the `firewallRules` field:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: GCPCluster
metadata:
  name: my-cluster
spec:
  network:
    firewall:
      defaultRulesManagement: "Managed"
      firewallRules:
        - name: "custom-ingress-rule"
          description: "Allow SSH and custom application traffic"
          direction: "Ingress"
          priority: 1000
          allowed:
            - IPProtocol: "TCP"
              ports:
                - "22"
                - "8080"
                - "8443"
          sourceRanges:
            - "10.0.0.0/8"
            - "172.16.0.0/12"
          targetTags:
            - "web-servers"
```

### FirewallRule Fields

| Field | Type | Required | Description |
|-------|------|----------|-------------|

> | `name` | string | Optional | Rule name (1-63 chars, must match `[a-z]([-a-z0-9]*[a-z0-9])?`). If not prefixed with cluster name, it will be prepended automatically. |

Reviewer: The doc defines name restrictions so but API doesn't enforce it. Should we?

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/95b916103f0b609a8556b2f3345325fbaa04771c/api/v1beta1/types.go#L207-L208

| `description` | string | Optional | Description of the rule (max 2000 chars). Defaults to "Created by Cluster API GCP Provider". |
| `direction` | string | Optional | Traffic direction: `Ingress` (default) or `Egress`. |

> | `priority` | integer | Optional | Rule priority (1-65535). Lower values = higher priority. Default: 1000. |

Reviewer: This `1-65535` is the restriction that comes from `GCPCluster` CEL.

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/95b916103f0b609a8556b2f3345325fbaa04771c/api/v1beta1/types.go#L218-L221

However, GCP allows `Integer from 0 to 65535` according https://docs.cloud.google.com/firewall/docs/firewalls#priority_order_for_firewall_rules.

Is this a CEL bug 👀?

| `allowed` | []FirewallDescriptor | Optional | List of ALLOW rules (max 1024). |

> | `denied` | []FirewallDescriptor | Optional | List of DENY rules (max 1024). DENY rules take precedence over ALLOW rules if they have equal priority. |

Reviewer: Field `allowed` and `denied` cannot be specified in a single rule, right? That is: a rule must either allow or deny (can't do both). We should mention it...

https://docs.cloud.google.com/firewall/docs/firewalls#action_of_the_rule

> | `sourceRanges` | []string | Optional | Source IP ranges in CIDR format (max 1024). Supports IPv4 and IPv6. |

Reviewer: `sourceRanges` is also _not_ applicable to `Egress` firewall rule, right?

| `sourceTags` | []string | Optional | Source instance tags (max 30, 1-63 chars each). Only applies to traffic between instances in the same VPC. Not valid for Egress rules. |

> | `destinationRanges` | []string | Optional | Destination IP ranges in CIDR format (max 1024). Only valid for Egress rules. |

Reviewer: These 3 fields are only applicable to either Ingress or Egress direction (i.e. not both). AFAICT, the GCP API will return `400` error if used in unsupported direction.

How about adding some CEL or webhook validation to block such case early? This way, in `GCPCluster` object won't get reconciled and avoid making invalid API calls.
